gavel_ci/templates/console.html

```diff
@@ -0,0 +1,89 @@
+{% extends "layout.html" %}
+{% block body %}
+<h1>Gavel CI Build: <a href="{{url_for('jobserv.project', proj=proj)}}">{{proj}}</a> <a href="{{url_for('jobserv.build',proj=proj, build=build)}}">#{{build}}</a> - {{run}}</h1>
+<h3>
+    console.log
+    <span id='console-status'></span>
+    <span id='console-raw' style="display:none;">
+        (<a 
+          href="{{url_for('jobserv.run_artifact', proj=proj, build=build, run=run, p='console.log')}}"
+          download="{{proj}}-{{build}}-{{run}}-console.log">raw</a>)
+    </span>
+</h3>
+<div>
+    <pre
+        id="console-log-tail"
+        style="white-space: pre-wrap;"
+    ></pre>
+</div>
+{% endblock %}
+{% block script %}
+{{super()}}
+<script type="text/javascript">
+(function () {
+    let tailElement = $("#console-log-tail");
+    let downloadElement = $("#console-raw");
+    let statusElement = $("#console-status");
+    var offset = 0;
+    let messageTailWindow = 40;
+    var messageTail = [];
+    var status = '';
+    var interval = 1000; /* 1s */
+    let url = "/projects/{{proj}}/builds/{{build}}/{{run}}/console/tail";
+    var showStatus = function (slug) {
+        let status = {
+            running: 'tailing',
+            queued: 'starting',
+        }[slug.toLowerCase()] || 'tail';
+        $(statusElement).text(status);
+    };
+    var showDownload = function () {
+        $(downloadElement).show();
+    }
+    var updateLog = function (message) {
+        if (!message) return;
+        const messageRows = message.split('\n');
+        messageTail = messageTail.concat(messageRows);
+        messageTail = messageTail.slice(Math.max(messageTail.length - messageTailWindow, 0));
+        let messageContent = messageTail.join('\n');
+        $(tailElement).text(`...\n${messageContent}`);
+    };
+    var rawGet = function (url) {
+        $.ajax(url, {
+            dataType: "text",
+            cache: false,
+            headers: {"X-OFFSET": `${offset}`, Range: "bytes=-2048"},
+            success: function (data, s, xhr) {
+                var newStatus = xhr.getResponseHeader('X-RUN-STATUS');
+                var newOffset = xhr.getResponseHeader('X-OFFSET');
+                if (newStatus == 'QUEUED') {
+                    if (!status) { // started
+            			updateLog(data);
+            		} else {
+            			updateLog(".");
+            		}
+                } else if (!newStatus) { // finished
+                    showDownload();
+                    updateLog(data);
+                    return;
+                } else {
+                    if (newStatus != status) {
+                        updateLog(`--- Status change: ${status} -> ${newStatus}`);
+			        }
+                    showStatus(newStatus);
+        			updateLog(data);
+                }
```

Comment:
looks like there might be some tabs/spaces thing going on with the indention here.

Reply:
uhh.. that is odd. I wonder if it's just this view?  Here's how it looks in my editor 
<img width="537" alt="image" src="https://user-images.githubusercontent.com/5916490/81228646-0476c800-8fb4-11ea-92a6-281f9bda6697.png">

Reply:
it was tabs -- good catch